packages/eslint-plugin/src/rules/no-confusing-void-expression.ts

```diff
@@ -388,5 +415,56 @@ export default createRule<Options, MessageId>({
       const type = getConstrainedTypeAtLocation(services, targetNode);
       return tsutils.isTypeFlagSet(type, ts.TypeFlags.VoidLike);
     }
+
+    function isFunctionReturnTypeIncludesVoid(functionType: ts.Type): boolean {
+      const callSignatures = tsutils.getCallSignaturesOfType(functionType);
+
+      return callSignatures.some(signature => {
+        const returnType = signature.getReturnType();
+
+        return tsutils
+          .unionTypeParts(returnType)
+          .some(tsutils.isIntrinsicVoidType);
+      });
+    }
+
+    function isVoidReturningFunctionNode(
+      functionNode:
+        | TSESTree.ArrowFunctionExpression
+        | TSESTree.FunctionDeclaration
+        | TSESTree.FunctionExpression,
+    ): boolean {
+      // Game plan:
+      //   - If the function node has a type annotation, check if it includes `void`.
+      //     - If it does then the function is safe to return `void` expressions in.
+      //   - Otherwise, check if the function is a function-expression or an arrow-function.
+      //   -   If it is, get its contextual type and bail if we cannot.
+      //   - Return based on whether the contextual type includes `void` or not
+
+      const functionTSNode = services.esTreeNodeToTSNodeMap.get(functionNode);
+
+      if (functionTSNode.type) {
+        const returnType = checker.getTypeFromTypeNode(functionTSNode.type);
+
+        return tsutils
+          .unionTypeParts(returnType)
+          .some(tsutils.isIntrinsicVoidType);
```

Comment:
[Testing] No tests fail if this is changed to `.every`. I think that means there's some logic being missed?